main.go

```diff
@@ -0,0 +1,517 @@
+	// "selector expression". Fortunately, these can be unwrapped to reveal
+	// the underlying `*ast.Ident`.
+	unwrap = func(t ast.Expr) *ast.Ident {
+		switch x := t.(type) {
+		case *ast.SelectorExpr:
+			return unwrap(x.Sel)
+		case *ast.StarExpr:
+			return unwrap(x.X)
+		case *ast.Ident:
+			return x
+		}
+		return nil
+	}
+
+	if len(f.Names) > 0 {
+		return p.TypesInfo.Defs[f.Names[0]].Type()
+	} else {
+		return p.TypesInfo.Defs[unwrap(f.Type)].Type()
+	}
+}
+
+func Type_IsArray(t types.Type) bool {
+	// In most circumstances the "intermediate" types matter. However,
+	// `toml` automatically constructs the intermediates when parsing.
+	// For instance, if a struct contains a field of type `*string`,
+	// `toml` can initialize it from a plain-old "string". We will
+	// ignore the intermediates when appropriate.
+	switch x := t.(type) {
+	case *types.Pointer:
+		return Type_IsArray(x.Elem())
+	case *types.Slice:
+		return true
+	}
+	return false
+}
+
+func Type_IsStruct(t types.Type) bool {
+	x := Type_GetBasal(t)
+
+	if _, ok := x.(*types.Struct); ok {
+		return true
+	}
+	if _, ok := x.Underlying().(*types.Struct); ok {
+		return true
+	}
+	return false
+}
+
+func Type_GetBasal(t types.Type) types.Type {
+	// In most circumstances the "intermediate" types matter. However,
+	// `toml` automatically constructs the intermediates when parsing.
+	// For instance, if a struct contains a field of type `*string`,
+	// `toml` can initialize it from a plain-old "string". We will
+	// ignore the intermediates when appropriate.
+	switch x := t.(type) {
+	case *types.Pointer:
+		return Type_GetBasal(x.Elem())
+	case *types.Slice:
+		return Type_GetBasal(x.Elem())
+	}
+	return t
+}
+
+func Type_GetNamed(t types.Type) *types.Named {
+	x, ok := Type_GetBasal(t).(*types.Named)
+	if !ok {
+		return nil
+	}
+	return x
+}
+
+func Type_LoadPackage(pkgs *map[string]Package, t *types.Named) *Package {
+	path := t.Obj().Pkg().Path()
+
+	if p, ok := (*pkgs)[path]; ok {
+		return &p
+	}
+
+	config := loader.Config{
+		ParserMode: parser.ParseComments,
+	}
+	config.Import(path)
+	prog, err := config.Load()
+	if err != nil {
+		panic(err)
+	}
+
+	for _, pkgInfo := range prog.InitialPackages() {
+		(*pkgs)[path] = Package{
+			Fset:      prog.Fset,
+			Files:     pkgInfo.Files,
+			Path:      pkgInfo.Pkg.Path(),
+			TypesInfo: &pkgInfo.Info,
+		}
+	}
+
+	if p, ok := (*pkgs)[path]; ok {
+		return &p
+	}
+
+	return nil
+}
+
+func WriteIndent(context *Context) {
+	for i := 0; i < context.Indent; i++ {
+		fmt.Printf("  ")
+	}
+}
+
+func WriteFieldStruct(context *Context, f *ast.Field) {
+	t := Field_GetType(context.Package, f)
+
+	n := Type_GetNamed(t)
+	if n == nil {
+		return
+	}
+
+	x := Type_LoadPackage(&context.Packages, n)
+	if x == nil {
+		return
+	}
+
+	ProcessStruct(context.StartStruct(x), n.Obj().Name())
+}
+
+func WriteComment(c *Context, g *ast.CommentGroup) {
+	// comment regex
+	rc := regexp.MustCompile("^\\/\\/(.*)")
```

Comment:
If performance is a concern, these could be moved to the package level and only compiled once.